cuda.core v1.0.0 release prep: support policy, release notes, doc cleanup (#2032)
NVIDIA/cuda-python: cuda_core/docs/source/support.rst

Author: Keith: We will drop 1.x as soon as 2.0 (aka CUDA 14) is out.

Author: aka project lifetime

Author: done in commit 6eb0a457216808634b2066aa4d9fb2e0ef1eaaf7

.. SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

.. _cuda-core-support:

``cuda.core`` Support Policy
============================

Versioning Scheme
-----------------

``cuda.core`` follows `Semantic Versioning (SemVer) <https://semver.org/>`_ with the version
format ``major.minor.patch``:

- **Major**: Bumped when a new CUDA major release is out and support for the oldest CUDA major
  version is dropped. Breaking API changes only happen at major-version boundaries.

> - **Minor**: Bumped when new, backward-compatible features are added, or when a new Python minor

Reviewer: Python calls these "feature releases", because "minor" sort of makes sense because it's the second number, it undersells the importance of them (and it doesn't follow semver anyway).

```suggestion
- **Minor**: Bumped when new, backward-compatible features are added, or when a new Python feature
```

  release is out and the oldest supported Python version reaches EOL.
- **Patch**: Bumped for bug fixes and backward-compatible maintenance updates.

Unlike ``cuda.bindings``, the ``cuda.core`` version is *not* aligned with the CUDA Toolkit version.
Consult the table below or the :doc:`release notes <release>` to determine which CUDA versions are
supported by a given ``cuda.core`` release.

CUDA Version Support
--------------------

``cuda.core`` is actively maintained to support the two (2) most recent CUDA major versions. For
example, ``cuda.core`` 1.x supports CUDA 12 and 13. Any fix in the latest release would be
backported as needed.

When a new CUDA major version is released and support for the oldest major version is dropped,
``cuda.core`` will release a new major version (e.g., 1.x → 2.0.0).

.. list-table:: CUDA Version Support Matrix
   :header-rows: 1

   * - ``cuda.core`` version
     - Supported CUDA versions
   * - 1.x
     - 12, 13

> 

Reviewer: Do we want to add that certain features in `cuda_core` may require a specific minimum version of `cuda_bindings`?  (I'm thinking of NVML support specifically, but that may not be the only case).  Saying it supports "any 12 or 13" is a bit misleading.

As a minimum, I'm going to file another pull request to document the minimum required cuda_bindings version that will set next to the `cuda.core.system` docs.

Author: Yes. I think so. We should note that as in any CUDA libraries, certain features may impose extra requirements on minimal cuda-bindings or CUDA driver versions.

Python Version Support
----------------------

``cuda.core`` supports all Python versions following the `CPython EOL schedule
<https://devguide.python.org/versions/>`_. As of writing, Python 3.10 – 3.14 are supported.

> When a new Python minor version is released and the oldest supported version reaches EOL,

Reviewer: ```suggestion
When a new Python feature version is released and the oldest supported version reaches EOL,
```

.. SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

.. _cuda-core-support:

``cuda.core`` Support Policy
============================

Versioning Scheme
-----------------

``cuda.core`` follows `Semantic Versioning (SemVer) <https://semver.org/>`_ with the version
format ``major.minor.patch``:

- **Major**: Bumped when a new CUDA major release is out and support for the oldest CUDA major
  version is dropped. Breaking API changes only happen at major-version boundaries.
- **Minor**: Bumped when new, backward-compatible features are added, or when a new Python feature
  release is out and the oldest supported Python version reaches EOL.
- **Patch**: Bumped for bug fixes and backward-compatible maintenance updates.

Unlike ``cuda.bindings``, the ``cuda.core`` version is *not* aligned with the CUDA Toolkit version.
Consult the table below or the :doc:`release notes <release>` to determine which CUDA versions are
supported by a given ``cuda.core`` release.

CUDA Version Support
--------------------

> ``cuda.core`` is actively maintained to support the two (2) most recent CUDA major versions. For
> example, ``cuda.core`` 1.x supports CUDA 12 and 13.
> 

Author: ```suggestion
``cuda.core`` is actively maintained to support the two (2) most recent CUDA major versions. For
example, ``cuda.core`` 1.x supports CUDA 12 and 13.

In particular, what this entails is that all CUDA minor versions within the two major versions (12.x, 13.x) are supported.
```

Author: done in commit 6eb0a457216808634b2066aa4d9fb2e0ef1eaaf7

.. SPDX-License-Identifier: Apache-2.0

.. _cuda-core-support:

``cuda.core`` Support Policy
============================

Versioning Scheme
-----------------

``cuda.core`` follows `Semantic Versioning (SemVer) <https://semver.org/>`_ with the version
format ``major.minor.patch``:

- **Major**: Bumped when a new CUDA major release is out and support for the oldest CUDA major
  version is dropped. Breaking API changes only happen at major-version boundaries.
- **Minor**: Bumped when new, backward-compatible features are added, or when a new Python feature
  release is out and the oldest supported Python version reaches EOL.
- **Patch**: Bumped for bug fixes and backward-compatible maintenance updates.

Unlike ``cuda.bindings``, the ``cuda.core`` version is *not* aligned with the CUDA Toolkit version.
Consult the table below or the :doc:`release notes <release>` to determine which CUDA versions are
supported by a given ``cuda.core`` release.

CUDA Version Support
--------------------

``cuda.core`` is actively maintained to support the two (2) most recent CUDA major versions. For
example, ``cuda.core`` 1.x supports CUDA 12 and 13. Any fix in the latest release would be

> backported as needed.

Reviewer 2: The statement about backporting is a bit confusing here where it makes it sound like there's a different `cuda.core` package per CUDA major version. Unless you're implying we'll backport fixes from `cuda.core 2.x` to `cuda.core 1.x`?

Author: Good catch, Keith. I presented the version scheme and [cuda-bindings support policy](https://nvidia.github.io/cuda-python/cuda-bindings/latest/support.html), and it probably confused my bot.

Author: fixed in commit 096eb46

example, ``cuda.core`` 1.x supports CUDA 12 and 13.

In particular, what this entails is that all CUDA minor versions within the two major releases
(12.x, 13.x) are supported by the same ``cuda-core`` package.

When a new CUDA major version is released and support for the oldest major version is dropped,
``cuda.core`` will release a new major version (e.g., 1.x → 2.0.0).

.. list-table:: CUDA Version Support Matrix
   :header-rows: 1

   * - ``cuda.core`` version
     - Supported CUDA versions
   * - 1.x
     - 12, 13

As with any CUDA library, certain features may impose additional requirements on the minimum
``cuda-bindings``, CUDA library, or CUDA driver versions. Refer to the individual module
documentation for details.

Python Version Support
----------------------

``cuda.core`` supports all Python versions following the `CPython EOL schedule
<https://devguide.python.org/versions/>`_. As of writing, Python 3.10 – 3.14 are supported.

When a new Python feature version is released and the oldest supported version reaches EOL,
``cuda.core`` will bump its minor version accordingly.

Free-threading Build Support
----------------------------

> As of ``cuda.core`` 1.0.0, wheels for the `free-threaded interpreter

Author: ```suggestion
Starting ``cuda-core`` 0.4.0, packages for the `free-threaded interpreter
```
